File: lib/steps/s07_transcript_merge.py
import json
from pathlib import Path
from typing import Any, Dict, List

import srt

from core.step_base import StepBase


class StepTranscriptMerge(StepBase):
    step_id = "s07_transcript_merge"
    depends_on = ["s03_subtitle_detect", "s05_asr", "s06_ocr"]

    def run(self, workspace: Path, config: Dict[str, Any], job_state: Any) -> Dict[str, Any]:
        detect_info = job_state.get_step_output("s03_subtitle_detect") or {}
        mode = detect_info.get("mode")

        ocr_file = workspace / "s06_ocr.json"
        asr_file = workspace / "s05_asr.json"
        
        ocr_segments = []
        asr_segments = []
        
        if ocr_file.exists():
            with open(ocr_file, "r", encoding="utf-8") as f:
                ocr_segments = json.load(f)
        if asr_file.exists():
            with open(asr_file, "r", encoding="utf-8") as f:
                asr_segments = json.load(f)

        # Prioritize OCR segments if present, fallback to embedded sub or ASR
        if len(ocr_segments) > 0:
            merged_segments = ocr_segments
        elif mode == "embedded" and detect_info.get("embedded_sub"):
            sub_file = Path(detect_info["embedded_sub"])
            if sub_file.exists():
                with open(sub_file, "r", encoding="utf-8") as f:
                    sub_data = list(srt.parse(f.read()))
                for s in sub_data:
                    merged_segments.append({
                        "start": round(s.start.total_seconds(), 3),
                        "end": round(s.end.total_seconds(), 3),
                        "text": s.content.strip()
                    })
        else:
            merged_segments = asr_segments

        from utils.repetition_cleaner import RepetitionCleaner
        merged_segments = RepetitionCleaner.clean_segments(merged_segments)

        # Save final transcript
        out_file = workspace / "s07_transcript.json"
        with open(out_file, "w", encoding="utf-8") as f:
            json.dump(merged_segments, f, ensure_ascii=False, indent=2)

        return {
            "transcript_file": str(out_file),
            "segment_count": len(merged_segments)
        }

File: video_engine/lib/steps/s07_transcript_merge.dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as p;
import '../core/job_state.dart';
import '../core/step_base.dart';

class StepTranscriptMerge extends StepBase {
  @override
  String get stepId => 's07_transcript_merge';

  @override
  List<String> get dependsOn => const ['s05_asr', 's06_ocr'];

  @override
  List<String> get stepConfigKeys => const [];

  @override
  Future<Map<String, dynamic>> run(
    Directory workspace,
    Map<String, dynamic> config,
    JobState jobState,
  ) async {
    final outFile = File(p.join(workspace.path, 's07_transcript.json'));

    final asrInfo = jobState.getStepOutput('s05_asr') ?? {};
    final ocrInfo = jobState.getStepOutput('s06_ocr') ?? {};

    final asrFile = asrInfo['transcript_file'] as String?;
    final ocrFile = ocrInfo['ocr_file'] as String?;

    List<dynamic> asrSegments = [];
    List<dynamic> ocrSegments = [];

    if (asrFile != null && File(asrFile).existsSync()) {
      try {
        asrSegments = jsonDecode(File(asrFile).readAsStringSync()) as List? ?? [];
      } catch (_) {}
    }

    if (ocrFile != null && File(ocrFile).existsSync()) {
      try {
        ocrSegments = jsonDecode(File(ocrFile).readAsStringSync()) as List? ?? [];
      } catch (_) {}
    }

    final isOcrOnly = config['ocr_only'] == true;
    List<Map<String, dynamic>> finalSegments = [];

    if (isOcrOnly || asrSegments.isEmpty) {
      finalSegments = ocrSegments.map((e) => Map<String, dynamic>.from(e as Map)).toList();
    } else {
      // Primary: ASR segments
      finalSegments = asrSegments.map((e) => Map<String, dynamic>.from(e as Map)).toList();
    }

    // 1. Sort by start time, then by longer duration
    finalSegments.sort((a, b) {
      final sa = double.tryParse(a['start']?.toString() ?? '0.0') ?? 0.0;
      final sb = double.tryParse(b['start']?.toString() ?? '0.0') ?? 0.0;
      if (sa != sb) return sa.compareTo(sb);
      final ea = double.tryParse(a['end']?.toString() ?? '0.0') ?? 0.0;
      final eb = double.tryParse(b['end']?.toString() ?? '0.0') ?? 0.0;
      return eb.compareTo(ea);
    });

    // 2. Smart Deduplication & Overlap Cleaning
    final cleanedSegments = _cleanAndDeduplicate(finalSegments);

    // 3. Re-index segment IDs
    for (int i = 0; i < cleanedSegments.length; i++) {
      cleanedSegments[i]['id'] = i;
    }

    outFile.writeAsStringSync(
      const JsonEncoder.withIndent('  ').convert(cleanedSegments),
      flush: true,
    );

    return {
      'transcript_file': outFile.path,
      'segment_count': cleanedSegments.length,
    };
  }

  List<Map<String, dynamic>> _cleanAndDeduplicate(List<Map<String, dynamic>> list) {
    if (list.isEmpty) return [];

    final result = <Map<String, dynamic>>[];

    for (int i = 0; i < list.length; i++) {
      final cur = Map<String, dynamic>.from(list[i]);
      final curText = (cur['text'] ?? '').toString().trim();
      if (curText.isEmpty) continue;

      double curStart = double.tryParse(cur['start']?.toString() ?? '0.0') ?? 0.0;
      double curEnd = double.tryParse(cur['end']?.toString() ?? '${curStart + 1.0}') ?? (curStart + 1.0);

      if (result.isNotEmpty) {
        final prev = result.last;
        final prevText = (prev['text'] ?? '').toString().trim();
        double prevStart = double.tryParse(prev['start']?.toString() ?? '0.0') ?? 0.0;
        double prevEnd = double.tryParse(prev['end']?.toString() ?? '${prevStart + 1.0}') ?? (prevStart + 1.0);

        final normCur = curText.toLowerCase().replaceAll(RegExp(r'\s+'), '');
        final normPrev = prevText.toLowerCase().replaceAll(RegExp(r'\s+'), '');

        // Case A: Exact duplicate text within close timestamp -> extend previous
        if (normCur == normPrev && (curStart <= prevEnd + 0.8)) {
          prev['end'] = curEnd > prevEnd ? curEnd : prevEnd;
          continue;
        }

        // Case B: Substring / Typing effect overlap (e.g. "Làm sao" then "Làm sao để biết", "等你长大了" then "等你长大了 山#")
        if ((normCur.contains(normPrev) || normPrev.contains(normCur)) && (curStart <= prevEnd + 0.6)) {
          if (curText.length > prevText.length) {
            prev['text'] = curText;
          }
          prev['end'] = curEnd > prevEnd ? curEnd : prevEnd;
          continue;
        }

        // Case C: Micro-noise filter (single-char fragments with short duration)
        if (curText.length <= 2 && (curEnd - curStart) < 0.6) {
          continue;
        }

        // Case D: Distinct segments overlap in time -> clamp prevEnd to avoid double sub flash
        if (prevEnd > curStart) {
          prev['end'] = curStart > prevStart + 0.3 ? curStart : (prevStart + 0.3);
        }
      }

      cur['start'] = double.parse(curStart.toStringAsFixed(3));
      cur['end'] = double.parse(curEnd.toStringAsFixed(3));
      result.add(cur);
    }

    return result;
  }
}
